### PR TITLE
Allow coordinate selection in calcdervis

### PR DESCRIPTION
**Pull request by jubich**
This PR adds the ability to specify the direction(s) when using the calcderivis file.

### CODE REVIEW

**Comment by bhourahine** [comment]
What is the use case for these changes? Also would it be useful to be able to specify specific atoms as well for derivatives (which is what I'd initially assumed was the functionality of this PR).

**Comment by jubich** [comment]
This should reduce the computation time for specific calculations where only the forces in certain directions are important. Also, it can be used for development. 

The point with the specific selection of certain atoms for the calculation will also be added with the next commit to this PR.

File: utils/numcheck/calcderivs [hunk 1]
```diff
@@ -73,18 +75,19 @@ def main():
 
     if calcforces:
         forces = calculate_forces( binary, disp, coords, specienames,
```
Comment (vanderhe) [inline utils/numcheck/calcderivs:77]:
```suggestion
        forces = calculate_forces(binary, disp, coords, specienames,
```

File: utils/numcheck/calcderivs [hunk 2]
```diff
@@ -73,18 +75,19 @@ def main():
...
-                                   species, origin, latvecs, chargeRestart)
+                                   species, origin, latvecs, chargeRestart,
+                                   direction, atoms)
```
Comment (vanderhe) [inline utils/numcheck/calcderivs:79]:
```suggestion
                                  species, origin, latvecs, chargeRestart,
                                  direction, atoms)
```

File: utils/numcheck/calcderivs [hunk 3]
```diff
@@ -73,18 +75,19 @@ def main():
...
 
     if latvecs is not None and calclatderivs:
         latderivs = calculate_latderivs( binary, disp, coords, latvecs,
                                          specienames, species, origin,
```
Comment (vanderhe) [inline utils/numcheck/calcderivs:83]:
```suggestion
        latderivs = calculate_latderivs(binary, disp, coords, latvecs,
                                        specienames, species, origin,
```

File: utils/numcheck/calcderivs [hunk 4]
```diff
@@ -73,18 +75,19 @@ def main():
...
-                                         chargeRestart)
+                                         chargeRestart, direction)
```
Comment (vanderhe) [inline utils/numcheck/calcderivs:84]:
```suggestion
                                        chargeRestart, direction)
```

File: utils/numcheck/calcderivs [hunk 5]
```diff
@@ -117,6 +120,17 @@ def parse_arguments():
...
                         action="store_true", default=False, help=msg)
 
+    msg = "Direction of calculation (arguments like 'xyz' 'xy', etc."\
+          "input order determines output order, default: xyz"
+    parser.add_argument("--direction", dest="direction", default="xyz",
+                        help=msg)
+    msg = "Atoms for calculation, stating with one and including both limits,"\
+          " use ':' for ranges, ',' for seperation and '1:-x' for excluding"\
+          " the last x atoms. Output order will be sorted smallest to "\
+          "highest (arguments like '1,2' '1:5', '2,4:6,9' and '1,3:4,9:-4')"
```
Comment (vanderhe) [inline utils/numcheck/calcderivs:130]:
```suggestion
    msg = "Atoms for calculation, starting with one and including both limits,"\
          " use ':' for ranges, ',' for separation and '1:-x' for excluding"\
          " the last x atoms. Output order will be sorted smallest to "\
          "highest (arguments like '1,2' '1:5', '2,4:6,9' and '1,3:4,9:-4')"
```

File: utils/numcheck/calcderivs [hunk 6]
```diff
@@ -168,18 +182,18 @@ def calculate_reference(binary, reffile, coords, specienames, species, origin,
 
 
 def calculate_forces(binary, disp, coords, specienames, species, origin,
-                     latvecs, chargeRestart):
+                     latvecs, chargeRestart, direction, atoms):
```
Comment (vanderhe) [inline utils/numcheck/calcderivs:185]: I think we should stick to PEP 8 for naming variables (avoid mixedCase), not our Fortran conventions (lowerCamelCase):
https://peps.python.org/pep-0008/#function-and-variable-names

**Review by vanderhe — COMMENTED** [review]
Some suggestions regarding style and formatting.

File: utils/numcheck/calcderivs [hunk 7]
```diff
@@ -233,41 +248,97 @@ def calculate_latderivs(binary, disp, coords, latvecs, specienames, species,
...
+def print_forces(forces, forces0, direction, atoms):
     """Prints calculates forces"""
 
-    print("Forces by finite differences:")
-    for iat, atforce in enumerate(forces):
-        print(("%25.12E" * 3) % tuple(atforce))
+    cart = ('x', 'y', 'z')
+    dir_str = ("%s" * len(direction)) % tuple(cart[x] for x in direction)
+    num = len(direction)
+    print("Lattice derivatives by finite differences ("+dir_str+"):")
```
Comment (vanderhe) [inline utils/numcheck/calcderivs:257]:
```suggestion
    print("Forces by finite differences ("+dir_str+"):")
```
Copy-paste error?

**Review by vanderhe — APPROVED** [review]
Ready to go from my side a.s.a. typo is corrected. Nice work, I actually already made use of it.

File: utils/numcheck/calcderivs (hunk 7 above)
Comment (jubich) [reply utils/numcheck/calcderivs]: Yes, looks very much like it

File: utils/numcheck/calcderivs [hunk 8]
```diff
@@ -233,41 +248,97 @@ def calculate_latderivs(binary, disp, coords, latvecs, specienames, species,
...
+        if coords.count(coord) > 1:
+            raise ValueError(f"'{coord}’ is specified more than once!")
+        if coord == "x":
+            direction.append(0)
+        elif coord == "y":
+            direction.append(1)
+        elif coord == "z":
+            direction.append(2)
+        else:
+            raise ValueError(f"'{coord}’ is no valid direction!")
```
Comment (aradi) [inline utils/numcheck/calcderivs:316]: I think, you can make it more elegant and compact:
```
dirstrs = {'x': 0, 'y': 1, 'z': 2}
directions = set()
for coord in coords:
    dir = distr.get(coord)
    if dir is None:
       raise ...
    directions.add(dir)
directions = sorted(d)
```

File: utils/numcheck/calcderivs [hunk 9]
```diff
@@ -233,41 +248,97 @@ def calculate_latderivs(binary, disp, coords, latvecs, specienames, species,
...
-        diff = latderivs - latderivs0
+        diff = latderivs - latderivs1
         for ii in range(3):
-            print(("%25.12E"*3) % tuple(diff[ii]))
+            print(("%25.12E" * num) % tuple(diff[ii]))
         print("Max diff in any lattice derivatives component:")
         print("%25.12E" % (abs(diff).max(), ))
 
 
+def direction_from_args(args):
```
Comment (aradi) [inline utils/numcheck/calcderivs:300]: pass in `args.direction` only, not `args`, as only `direction` is used.

File: utils/numcheck/calcderivs [hunk 10]
```diff
@@ -233,41 +248,97 @@ def calculate_latderivs(binary, disp, coords, latvecs, specienames, species,
...
+                else:
+                    start, stop = item.split(":")
+                for num in range(int(start), int(stop)+1):
+                    atoms.append(num-1)
+            else:
+                atoms.append(int(item)-1)
+        atoms.sort()
+    else:
+        for num in range(0,maxlen):
+            atoms.append(num)
```
Comment (aradi) [inline utils/numcheck/calcderivs:340]: `atoms = list(range(maxlen))`

File: utils/numcheck/calcderivs [hunk 11]
```diff
@@ -233,41 +248,97 @@ def calculate_latderivs(binary, disp, coords, latvecs, specienames, species,
...
+        elif coord == "y":
+            direction.append(1)
+        elif coord == "z":
+            direction.append(2)
+        else:
+            raise ValueError(f"'{coord}’ is no valid direction!")
+    return direction
+
+
+def atoms_from_args(args, maxlen):
```
Comment (aradi) [inline utils/numcheck/calcderivs:320]: Pass in `args.atoms` only, `args` itself is not used anywhere.

File: utils/numcheck/calcderivs [hunk 12]
```diff
@@ -157,73 +171,74 @@ def read_reference_results(autotest0, calcforces, calclatderivs, latvecs):
...
+                        origin, charge_restart, direction):
     """Calculates lattice derivatives by finite differences"""
 
+    cart =('x','y','z')
     delta =('-','+')
     energy = np.empty((2,), dtype=float)
-    latderivs = np.empty((3, 3), dtype=float)
-    for ii in range(3):
+    latderivs = np.empty((3, len(direction)), dtype=float)
+    for ii, coord in enumerate(direction):
```
Comment (aradi) [inline utils/numcheck/calcderivs:224]: `ii` is not used inside the loop

**Review by aradi — CHANGES_REQUESTED** [review]

**Comment by aradi** [comment]
One more comment: `direction` should be called `directions` as it is a list of (eventually) multiple entries.

File: utils/numcheck/calcderivs (hunk 12 above)
Comment (jubich) [reply utils/numcheck/calcderivs:224]: It is used at the end of that loop for filling into `latderivs`.

https://github.com/dftbplus/dftbplus/blob/d20a9e5d9e96274ac95134ee981dc0242058b349/utils/numcheck/calcderivs#L245-L248

Comment (aradi) [reply utils/numcheck/calcderivs]: Sorry, I've missed that.

**Review by aradi — APPROVED** [review]

**Pull request merged**